# Numpy
package yang digunakan untuk handling data tipe numerik

In [ ]:
import numpy as np
import pandas as pd # untuk pandas

In [ ]:
# perbedaan list dan numpy array, beda data type
list_angka = [50,60,70,80]
arr_angka = np.array(list_angka)

print(list_angka, type(list_angka))
print(arr_angka, type(arr_angka))

[50, 60, 70, 80] <class 'list'>
[50 60 70 80] <class 'numpy.ndarray'>


In [ ]:
# aritmatik
# misalkan kita mau tambahkan 100 ke semua angka di dalam list_angka
index = 0
for n in list_angka:
  list_angka[index] += 100
  index += 1

print(list_angka)

[150, 160, 170, 180]


In [ ]:
arr_angka + 100

array([150, 160, 170, 180])

untuk melakukan aritmatik ke banyak data sekaligus, numpy array terbaik, mudah, simpel. perhatikan bagaimana jauh lebih sulit melakukan aritmatik dengan list, walaupun isinya sudah angka semua.

In [ ]:
# data 2 dimensi
# ada baris, ada kolom
arr_penjualan = np.array([
    [10, 20, 11],
    [50, 1, 0],
])

print(arr_penjualan)

[[10 20 11]
 [50  1  0]]


berbeda dengan list yang menderetkan semua datanya ke kanan, numpy array juga bisa memperlihatkan data tipe 2 dimensi dalam bentuk baris dan kolom

kekurangan numpy array adalah dia masih menggunakan index angka untuk mengakses item di dalam. belum punya judul baris dan kolom yang bisa kita pakai - itu nanti di pandas di sesi berikut.

## Numpy Array
CRUD - create read update delete


In [ ]:
# Create
# kita bisa buat numpy array dari list angka yang kita konversi menjadi array
# kita bisa juga generate angka untuk mengisi sebuah numpy array ukuran tertentu

# anggap kita mau membuat numpy array nilai nilai murid di kelas
arr_nilai = np.array([100, 70, 98, 57, 65, 84, 23])
print(arr_nilai)

[100  70  98  57  65  84  23]


In [ ]:
arr_nilai2 = np.array([[100, 70, 98, 57, 65, 84, 23], [89,54,78,56,57,98,12]])
print(arr_nilai2)

[[100  70  98  57  65  84  23]
 [ 89  54  78  56  57  98  12]]


In [ ]:
arr_nilai = np.ones(7)
print(arr_nilai)

[1. 1. 1. 1. 1. 1. 1.]


In [ ]:
arr_nilai2 = np.ones((7,3))
print(arr_nilai2)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [ ]:
arr_nilai = np.zeros(10)
print(arr_nilai)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [ ]:
arr_nilai2 = np.zeros((3,10))
print(arr_nilai2)

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]


In [ ]:
arr_nilai = np.random.random(7)
print(arr_nilai)

[0.28031311 0.95264035 0.26062939 0.10168596 0.12969986 0.89494952
 0.02321754]


In [ ]:
arr_nilai2 = np.random.random((5,6))
print(arr_nilai2)

[[0.00548634 0.55365949 0.25373214 0.69051891 0.12642629 0.56881457]
 [0.56349147 0.88669409 0.80448277 0.9887717  0.61701083 0.25849694]
 [0.32605477 0.57212267 0.95320002 0.62771033 0.17734715 0.94370469]
 [0.47667896 0.6594512  0.8687426  0.07531962 0.57256632 0.21784224]
 [0.36602731 0.86039852 0.48084738 0.82585863 0.79304441 0.50838157]]


In [ ]:
arr_nilai = np.random.random(7) * 100
print(arr_nilai)

[34.33790139 19.57072088 89.21638019 60.37427106 19.3423971  14.639783
 61.04012685]


In [ ]:
arr_nilai2 = np.random.random((5,6)) * 100
print(arr_nilai2)

[[45.19676066 75.93409455 36.64792325 68.17457081 80.55970908 84.90946413]
 [31.83680737  3.68539889 11.45196468 29.13899985  1.8934157  13.87006455]
 [79.05868599 16.46994642 69.67032484 93.85542789 82.32432026 52.85749465]
 [60.89944217 79.80735264 12.64391426 79.98294022 81.00329415  1.09761148]
 [30.93737347 66.99501383 43.76251731 23.35553731 36.04865629 44.52915102]]


angka yang dibuat random random ada di antara 0 dan 1, jadi kalau kita mau buat angka antara 0-100, tinggal kita kalikan 100.

untuk mengurangi desimal poin kita bisa gunakan teknik terpisah `np.round`

In [ ]:
arr_nilai = np.round(np.random.random(7) * 100, 2)
print(arr_nilai)

[26.24 45.9  94.23 73.41 48.03 44.1   6.77]


In [ ]:
arr_nilai2 = np.round(np.random.random((5, 10)) * 100, 2)
print(arr_nilai2)

[[40.44 52.48 63.65 41.15 10.11 11.83  8.35 83.81 48.04 46.94]
 [68.81 21.59 26.65 51.93 27.66 80.61 92.24 13.7   6.35 91.17]
 [19.14 49.77 11.24 19.19  1.69 70.15 19.26 20.87 26.65 69.13]
 [10.89 97.01 26.66 70.55 93.38 48.3  12.98 73.5  45.9  68.57]
 [50.24 60.98  9.54 46.09 25.41 45.3  45.11 84.55 58.05 64.46]]


In [ ]:
# kalau kita mau bikin range angka random yang berbeda, kita bisa pakai juga bantuan + - * / dan lainnya
# misalnya, kita mau buat hasil lempar dadu antara 1-6

arr_dadu = np.round(np.random.random((10, 10)) * 5 + 1, 0)
print(arr_dadu)

[[2. 5. 6. 2. 4. 3. 4. 2. 3. 3.]
 [6. 2. 5. 4. 6. 2. 6. 2. 3. 1.]
 [2. 3. 3. 5. 1. 2. 4. 2. 5. 2.]
 [2. 2. 5. 6. 4. 2. 3. 2. 3. 4.]
 [3. 3. 4. 5. 5. 5. 5. 1. 2. 4.]
 [1. 5. 5. 6. 5. 2. 6. 5. 5. 1.]
 [5. 4. 3. 2. 1. 1. 3. 4. 4. 6.]
 [6. 6. 5. 3. 1. 2. 4. 2. 2. 3.]
 [2. 4. 2. 5. 1. 4. 4. 4. 4. 3.]
 [4. 3. 3. 5. 4. 4. 6. 3. 5. 3.]]


In [ ]:
# reshape
print(arr_dadu.reshape((5, 20)))

# jumlah item sebelum dan sesudah reshape harus sama

[[2. 5. 6. 2. 4. 3. 4. 2. 3. 3. 6. 2. 5. 4. 6. 2. 6. 2. 3. 1.]
 [2. 3. 3. 5. 1. 2. 4. 2. 5. 2. 2. 2. 5. 6. 4. 2. 3. 2. 3. 4.]
 [3. 3. 4. 5. 5. 5. 5. 1. 2. 4. 1. 5. 5. 6. 5. 2. 6. 5. 5. 1.]
 [5. 4. 3. 2. 1. 1. 3. 4. 4. 6. 6. 6. 5. 3. 1. 2. 4. 2. 2. 3.]
 [2. 4. 2. 5. 1. 4. 4. 4. 4. 3. 4. 3. 3. 5. 4. 4. 6. 3. 5. 3.]]


In [ ]:
# kita juga bisa generate angka urut di numpy array
# cara pakainya sama dengan range

arr_urut = np.arange(10) # 0 - 9
print(arr_urut)

arr_urut = np.arange(10, 20) # 10 - 19
print(arr_urut)

arr_urut = np.arange(10, 100, 18) # 10, 28, 46, ...
print(arr_urut)

[0 1 2 3 4 5 6 7 8 9]
[10 11 12 13 14 15 16 17 18 19]
[10 28 46 64 82]


In [ ]:
arr_urut = np.arange(1900, 2025, 4) # 1900, 1904, 1908, ...
print(arr_urut)
print(arr_urut.shape)

[1900 1904 1908 1912 1916 1920 1924 1928 1932 1936 1940 1944 1948 1952
 1956 1960 1964 1968 1972 1976 1980 1984 1988 1992 1996 2000 2004 2008
 2012 2016 2020 2024]
(32,)


In [ ]:
print(arr_urut.reshape((4,8)))

[[1900 1904 1908 1912 1916 1920 1924 1928]
 [1932 1936 1940 1944 1948 1952 1956 1960]
 [1964 1968 1972 1976 1980 1984 1988 1992]
 [1996 2000 2004 2008 2012 2016 2020 2024]]


In [ ]:
print(np.arange(1900, 2025, 4).reshape((4,8)))

[[1900 1904 1908 1912 1916 1920 1924 1928]
 [1932 1936 1940 1944 1948 1952 1956 1960]
 [1964 1968 1972 1976 1980 1984 1988 1992]
 [1996 2000 2004 2008 2012 2016 2020 2024]]


ada beberapa cara untuk membuat numpy array
- np.array() diisi dengan list angka, bisa 1 dimensi, bisa 2 dimensi
- np.zeros() diisi dengan jumlah angka yang mau dibuat, nanti value nya 0 semua
- np.ones() diisi dengan jumlah angka yang mau dibuat, nanti value nya 1 semua
- np.random.random() diisi dengan jumlah angka yang mau dibuat, nanti value nya antara 0-1, float, punya desimal banyak
- np.arange() cara pakai sama dengan range, numpy array yang dibuat akan berisi angka urut


sebenarnya masih banyak lagi cara membuat numpy array, tapi jarang kita gunakan.



### latihan pendek (5 menit)
- buatlah sebuah numpy array menggunakan 5 cara berbeda diatas.
- cobalah untuk print numpy array yang sudah dibuat

## read
numpy array yang sudah dibuat, bisa kita lihat isinya

### shape
- ada berapa item/angka di dalam sebuah numpy array?
- ada berapa dimensi?
- setiap dimensi ada berapa item?

### indexing dan slicing
- angka ke-sekian di sebuah numpy array berapa?
- angka ke-2 sampai ke-5 di sebuah numpy array, berapa aja?

In [ ]:
arr_urut

array([10, 28, 46, 64, 82])

In [ ]:
print(arr_urut.size)
print(arr_urut.shape)
print(arr_urut.ndim)

5
(5,)
1


- ada 5 item/angka di dalam numpy array arr_urut
- di dimensi pertama arr_urut ada 5 item/angka
- arr_urut memiliki 1 dimensi

In [ ]:
arr_penjualan

array([[10, 20, 11],
       [50,  1,  0]])

In [ ]:
print(arr_penjualan.size)
print(arr_penjualan.shape)
print(arr_penjualan.ndim)

6
(2, 3)
2


- ada 6 item/angka di dalam numpy array arr_penjualan
- di dimensi pertama arr_penjualan ada 2 item/angka. di dimensi kedua arr_penjualan ada 3 item/angka.
- arr_penjualan memiliki 2 dimensi



### indexing dan slicing
- angka ke-sekian di sebuah numpy array berapa?
- angka ke-2 sampai ke-5 di sebuah numpy array, berapa aja?

In [ ]:
arr_urut

array([10, 28, 46, 64, 82])

In [ ]:
# angka pertama
print(arr_urut[0])

10

In [ ]:
# angka kedua dari depan
print(arr_urut[1])

28


In [ ]:
# angka terakhir
print(arr_urut[-1])

82


In [ ]:
# angka ke-2 sampai ke-5
print(arr_urut[1:5])

[28 46 64 82]


In [ ]:
# angka ke-2 dan ke-5
print(arr_urut[[1,4]])

[28 82]


demikian teknik indexing (ambil 1 item) dan slicing (ambil bbrp item)

In [ ]:
arr_penjualan

array([[10, 20, 11],
       [50,  1,  0]])

In [ ]:
print(arr_penjualan[0])

[10 20 11]


untuk numpy array 2 dimensi, saat kita mau ambil item, kita perlu 2 index juga. index untuk baris, dan index untuk kolom.

kalau cuma kasih 1 index, kita akan dapat baris, seluruh barisnya. baris pertama, kedua, ...

In [ ]:
# angka 10, di baris pertama, kolom pertama
print(arr_penjualan[0, 0])

10


In [ ]:
# angka 20, di baris pertama, kolom kedua
print(arr_penjualan[0, 1])

20


In [ ]:
# angka 1, di baris kedua, kolom kedua
print(arr_penjualan[1, 1])

1


In [ ]:
# angka 0, di baris terakhir, kolom terakhir
print(arr_penjualan[-1, -1])

# angka 0, di baris kedua, kolom ketiga
print(arr_penjualan[1, 2])

0
0


In [ ]:
arr_penjualan

array([[10, 20, 11],
       [50,  1,  0]])

In [ ]:
# ambil 10, 20, 50, 1
# baris 1 dan baris 2 --> 0:2
# kolom 1 dan kolom 2 --> 0:2
print(arr_penjualan[0:2, 0:2])

[[10 20]
 [50  1]]


In [ ]:
# ambil 10 dan 50, baris 1 dan 2, kolom 1
print(arr_penjualan[0:2, 0])

[10 50]


In [ ]:
# ambil 20 dan 11, baris 1, kolom 2 dan 3
print(arr_penjualan[0, 1:3])

[20 11]


Saat kita membuat numpy array, kita bisa membuat numpy array 2 dimensi juga. kita kembali ke atas untuk update code nya membuat versi 2 dimensi juga.

## reshape
numpy array yang udah jadi, bisa kita ubah bentuk shape nya. 1D jadi 2D, 3 baris 4 kolom jadi 4 baris 3 kolom, dll. bisa kita ubah bentuknya

In [ ]:
# reshape
arr_penjualan

array([[10, 20, 11],
       [50,  1,  0]])

In [ ]:
arr_penjualan.reshape((6,))

array([10, 20, 11, 50,  1,  0])

In [ ]:
arr_penjualan.reshape((3,2))

array([[10, 20],
       [11, 50],
       [ 1,  0]])

### latihan pendek (5 menit)
- cobalah buat numpy array 2 dimensi ukuran 14 * 10
- gunakan numpy random random untuk value antara 15 - 30
- gunakan reshape untuk mengubah bentuknya jadi 20 baris, 7 kolom

In [ ]:
arr_lat = np.ones((14, 10))
print(arr_lat)
print(arr_lat.size, arr_lat.shape)

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]
140 (14, 10)


In [ ]:
arr_lat = np.round(np.random.random((14, 10)) * 30 + 15, 1)
print(arr_lat)
print(arr_lat.size, arr_lat.shape)

[[31.1 25.9 25.5 33.3 32.2 15.3 29.1 16.3 35.1 43.6]
 [31.9 36.4 18.2 37.3 40.6 40.3 15.3 23.  20.7 21.9]
 [34.7 33.6 39.8 28.  34.1 19.7 33.9 31.  25.4 43.1]
 [32.1 42.6 38.  43.3 40.  44.4 44.6 27.9 26.3 32.2]
 [16.5 40.3 32.2 15.4 26.  35.3 41.1 27.7 35.  27.8]
 [30.7 42.5 18.4 25.9 42.7 37.9 17.8 32.4 41.  19.3]
 [32.5 21.1 16.7 33.4 16.3 36.4 15.  38.9 35.4 38.6]
 [18.  17.1 34.7 35.5 31.5 41.8 29.  36.6 18.9 17.2]
 [21.1 42.6 30.7 20.6 30.9 16.9 18.1 41.5 23.4 15.6]
 [35.6 37.3 32.7 32.7 40.6 40.1 24.1 17.2 21.2 43.9]
 [21.1 44.4 37.2 27.8 25.  39.6 43.1 21.8 42.1 17.6]
 [37.9 40.1 24.  22.6 23.4 16.8 39.2 28.6 36.8 43.2]
 [31.7 21.3 41.8 18.7 19.  23.9 16.7 40.7 38.4 39.6]
 [24.1 35.  43.1 24.6 26.8 44.5 25.7 24.1 28.2 27.7]]
140 (14, 10)


In [ ]:
print(arr_lat.reshape((20, 7)))

[[31.1 25.9 25.5 33.3 32.2 15.3 29.1]
 [16.3 35.1 43.6 31.9 36.4 18.2 37.3]
 [40.6 40.3 15.3 23.  20.7 21.9 34.7]
 [33.6 39.8 28.  34.1 19.7 33.9 31. ]
 [25.4 43.1 32.1 42.6 38.  43.3 40. ]
 [44.4 44.6 27.9 26.3 32.2 16.5 40.3]
 [32.2 15.4 26.  35.3 41.1 27.7 35. ]
 [27.8 30.7 42.5 18.4 25.9 42.7 37.9]
 [17.8 32.4 41.  19.3 32.5 21.1 16.7]
 [33.4 16.3 36.4 15.  38.9 35.4 38.6]
 [18.  17.1 34.7 35.5 31.5 41.8 29. ]
 [36.6 18.9 17.2 21.1 42.6 30.7 20.6]
 [30.9 16.9 18.1 41.5 23.4 15.6 35.6]
 [37.3 32.7 32.7 40.6 40.1 24.1 17.2]
 [21.2 43.9 21.1 44.4 37.2 27.8 25. ]
 [39.6 43.1 21.8 42.1 17.6 37.9 40.1]
 [24.  22.6 23.4 16.8 39.2 28.6 36.8]
 [43.2 31.7 21.3 41.8 18.7 19.  23.9]
 [16.7 40.7 38.4 39.6 24.1 35.  43.1]
 [24.6 26.8 44.5 25.7 24.1 28.2 27.7]]


sebelum di reshape ada 140 item/angka

kalau kita ubah ke bentuknya 20, 10 maka dia ekspektasi ada 200 item/angka

kalau tidak pas, jumlah item sebelum dan sesudah reshape, dia tidak bisa reshape

* jangan lupa reassign setelah reshape, kalau tidak hasil reshape tidak di save

* reshape tidak mengubah data awal

## update
- untuk mengubah angka tertentu di numpy array, kita bisa gunakan teknik indexing tadi, langsung reassign

 misalkan `arr_lat[0,0] = 100`

- untuk mengubah satu grup angka, bisa menggunakan teknik slicing

 misalkan `arr_lat[0:3,3] = 100` atau `arr_lat[:,3] = range(20)`

- untuk menambah numpy array bisa per baris atau per kolom (2D)

 kita pakai `np.concatenate()` untuk menempelkan 2 atau lebih numpy array jadi satu

In [ ]:
arr_lat[0,0] = 100
print(arr_lat)

[[100.   25.9  25.5  33.3  32.2  15.3  29.1  16.3  35.1  43.6]
 [ 31.9  36.4  18.2  37.3  40.6  40.3  15.3  23.   20.7  21.9]
 [ 34.7  33.6  39.8  28.   34.1  19.7  33.9  31.   25.4  43.1]
 [ 32.1  42.6  38.   43.3  40.   44.4  44.6  27.9  26.3  32.2]
 [ 16.5  40.3  32.2  15.4  26.   35.3  41.1  27.7  35.   27.8]
 [ 30.7  42.5  18.4  25.9  42.7  37.9  17.8  32.4  41.   19.3]
 [ 32.5  21.1  16.7  33.4  16.3  36.4  15.   38.9  35.4  38.6]
 [ 18.   17.1  34.7  35.5  31.5  41.8  29.   36.6  18.9  17.2]
 [ 21.1  42.6  30.7  20.6  30.9  16.9  18.1  41.5  23.4  15.6]
 [ 35.6  37.3  32.7  32.7  40.6  40.1  24.1  17.2  21.2  43.9]
 [ 21.1  44.4  37.2  27.8  25.   39.6  43.1  21.8  42.1  17.6]
 [ 37.9  40.1  24.   22.6  23.4  16.8  39.2  28.6  36.8  43.2]
 [ 31.7  21.3  41.8  18.7  19.   23.9  16.7  40.7  38.4  39.6]
 [ 24.1  35.   43.1  24.6  26.8  44.5  25.7  24.1  28.2  27.7]]


In [ ]:
arr_lat_reshaped = arr_lat.reshape((20,7))
arr_lat_reshaped[0:3,3] = 0
print(arr_lat_reshaped)

[[100.   25.9  25.5   0.   32.2  15.3  29.1]
 [ 16.3  35.1  43.6   0.   36.4  18.2  37.3]
 [ 40.6  40.3  15.3   0.   20.7  21.9  34.7]
 [ 33.6  39.8  28.   34.1  19.7  33.9  31. ]
 [ 25.4  43.1  32.1  42.6  38.   43.3  40. ]
 [ 44.4  44.6  27.9  26.3  32.2  16.5  40.3]
 [ 32.2  15.4  26.   35.3  41.1  27.7  35. ]
 [ 27.8  30.7  42.5  18.4  25.9  42.7  37.9]
 [ 17.8  32.4  41.   19.3  32.5  21.1  16.7]
 [ 33.4  16.3  36.4  15.   38.9  35.4  38.6]
 [ 18.   17.1  34.7  35.5  31.5  41.8  29. ]
 [ 36.6  18.9  17.2  21.1  42.6  30.7  20.6]
 [ 30.9  16.9  18.1  41.5  23.4  15.6  35.6]
 [ 37.3  32.7  32.7  40.6  40.1  24.1  17.2]
 [ 21.2  43.9  21.1  44.4  37.2  27.8  25. ]
 [ 39.6  43.1  21.8  42.1  17.6  37.9  40.1]
 [ 24.   22.6  23.4  16.8  39.2  28.6  36.8]
 [ 43.2  31.7  21.3  41.8  18.7  19.   23.9]
 [ 16.7  40.7  38.4  39.6  24.1  35.   43.1]
 [ 24.6  26.8  44.5  25.7  24.1  28.2  27.7]]


In [ ]:
arr_lat_reshaped[0:3,3] = [5,6,7]
print(arr_lat_reshaped)

[[100.   25.9  25.5   5.   32.2  15.3  29.1]
 [ 16.3  35.1  43.6   6.   36.4  18.2  37.3]
 [ 40.6  40.3  15.3   7.   20.7  21.9  34.7]
 [ 33.6  39.8  28.   34.1  19.7  33.9  31. ]
 [ 25.4  43.1  32.1  42.6  38.   43.3  40. ]
 [ 44.4  44.6  27.9  26.3  32.2  16.5  40.3]
 [ 32.2  15.4  26.   35.3  41.1  27.7  35. ]
 [ 27.8  30.7  42.5  18.4  25.9  42.7  37.9]
 [ 17.8  32.4  41.   19.3  32.5  21.1  16.7]
 [ 33.4  16.3  36.4  15.   38.9  35.4  38.6]
 [ 18.   17.1  34.7  35.5  31.5  41.8  29. ]
 [ 36.6  18.9  17.2  21.1  42.6  30.7  20.6]
 [ 30.9  16.9  18.1  41.5  23.4  15.6  35.6]
 [ 37.3  32.7  32.7  40.6  40.1  24.1  17.2]
 [ 21.2  43.9  21.1  44.4  37.2  27.8  25. ]
 [ 39.6  43.1  21.8  42.1  17.6  37.9  40.1]
 [ 24.   22.6  23.4  16.8  39.2  28.6  36.8]
 [ 43.2  31.7  21.3  41.8  18.7  19.   23.9]
 [ 16.7  40.7  38.4  39.6  24.1  35.   43.1]
 [ 24.6  26.8  44.5  25.7  24.1  28.2  27.7]]


In [ ]:
# ubah seluruh kolom ke-4
arr_lat_reshaped[:,3] = range(20)
print(arr_lat_reshaped)

[[100.   25.9  25.5   0.   32.2  15.3  29.1]
 [ 16.3  35.1  43.6   1.   36.4  18.2  37.3]
 [ 40.6  40.3  15.3   2.   20.7  21.9  34.7]
 [ 33.6  39.8  28.    3.   19.7  33.9  31. ]
 [ 25.4  43.1  32.1   4.   38.   43.3  40. ]
 [ 44.4  44.6  27.9   5.   32.2  16.5  40.3]
 [ 32.2  15.4  26.    6.   41.1  27.7  35. ]
 [ 27.8  30.7  42.5   7.   25.9  42.7  37.9]
 [ 17.8  32.4  41.    8.   32.5  21.1  16.7]
 [ 33.4  16.3  36.4   9.   38.9  35.4  38.6]
 [ 18.   17.1  34.7  10.   31.5  41.8  29. ]
 [ 36.6  18.9  17.2  11.   42.6  30.7  20.6]
 [ 30.9  16.9  18.1  12.   23.4  15.6  35.6]
 [ 37.3  32.7  32.7  13.   40.1  24.1  17.2]
 [ 21.2  43.9  21.1  14.   37.2  27.8  25. ]
 [ 39.6  43.1  21.8  15.   17.6  37.9  40.1]
 [ 24.   22.6  23.4  16.   39.2  28.6  36.8]
 [ 43.2  31.7  21.3  17.   18.7  19.   23.9]
 [ 16.7  40.7  38.4  18.   24.1  35.   43.1]
 [ 24.6  26.8  44.5  19.   24.1  28.2  27.7]]


In [ ]:
# untuk coba reshape, ukuran 24 enak karena bisa berubah jadi banyak bentuk berbeda
arr24 = np.ones(24)
print(arr24)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [ ]:
print(arr24.reshape((2,12)))

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


In [ ]:
print(arr24.reshape((3,8)))

[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]]


In [ ]:
print(arr24.reshape((4,6)))

[[1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]]


In [ ]:
print(arr24.reshape((6,4)))

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


`np.concatenate()` bisa kita gunakan untuk 'menempel' beberapa numpy array sekaligus

In [ ]:
arr1 = np.array([1,2,3])
arr2 = np.array([4,5,6])

np.concatenate((arr1, arr2))

array([1, 2, 3, 4, 5, 6])

In [ ]:
arr1 = np.array([[1,2,3],[4,5,6]])
arr2 = np.array([[0,8,9],[9,8,7]])

np.concatenate((arr1, arr2))

array([[1, 2, 3],
       [4, 5, 6],
       [0, 8, 9],
       [9, 8, 7]])

## Aritmatika di Numpy Array
selain membuat numpy array, dan mengedit numpy array, kita bisa melakukan aritmatika - tujuan utama package Numpy - dengan bantuan Numpy array

- rata-rata (mean)
- median
- sum
- max
- min
- tambah kurang kali bagi
- dan banyak lagi

In [ ]:
# anggap kita punya data temperature selama 20 minggu terakhir

arr_temp = np.round(np.random.random((20,7)) * 30 + 15, 1)
print(arr_temp)
print(arr_temp.size, arr_temp.shape)

[[23.  20.3 37.6 30.8 19.4 29.6 27.2]
 [25.9 30.4 43.2 16.8 42.6 28.  19.9]
 [42.9 31.1 17.7 40.1 34.9 20.7 21.3]
 [24.5 33.1 32.9 20.  34.8 29.6 38.6]
 [34.1 32.5 28.9 34.2 22.  18.1 38.7]
 [43.6 44.7 42.9 42.1 28.5 40.4 32.5]
 [16.3 41.3 25.1 16.5 29.2 32.  19.5]
 [25.  16.6 19.4 42.3 38.7 17.7 33.5]
 [41.3 44.  15.9 18.2 40.4 43.3 35.8]
 [17.1 19.8 37.5 36.7 35.  17.9 32.7]
 [33.1 18.9 18.  44.  20.  15.1 28.9]
 [25.3 38.1 37.4 31.1 19.7 39.6 29.4]
 [21.8 42.6 18.2 22.1 23.2 40.9 16. ]
 [38.5 38.5 41.1 15.9 36.9 35.1 36.5]
 [20.4 33.  19.7 16.8 36.8 40.2 20.2]
 [26.7 29.1 42.4 44.3 36.3 39.  27.7]
 [23.9 42.3 20.5 40.7 32.  38.  35.8]
 [41.7 38.2 34.8 18.8 18.6 30.8 21.2]
 [41.7 41.7 18.2 24.9 35.4 34.4 43.1]
 [30.5 22.  40.3 23.3 26.8 33.7 38.7]]
140 (20, 7)


- rata-rata harian temperatur 20 minggu terakhir berapa?
- per minggunya rata2nya berapa?
- max dan min temp selama 140 hari terakhir berapa?
- max dan min per minggu?

In [ ]:
arr_temp.mean()

30.42357142857143

rata-rata temperatur harian 20 minggu terakhir adalah 30.4 derajat

In [ ]:
mingguan = np.round(arr_temp.mean(axis=1), 1)
print(mingguan, mingguan.shape)

[26.8 29.5 29.8 30.5 29.8 39.2 25.7 27.6 34.1 28.1 25.4 31.5 26.4 34.6
 26.7 35.1 33.3 29.2 34.2 30.8] (20,)


kita punya data rata-rata temperatur per minggu selama 20 minggu terakhir

In [ ]:
# kalau rata2 temperatur hari senin berapa?
# rata2 1 kolom, untuk senin kolom pertama
weekly = np.round(arr_temp.mean(axis=0), 1)
print(weekly, weekly.shape)

[29.9 32.9 29.6 29.  30.6 31.2 29.9] (7,)


selain mean, kita bisa gunakan min, max, dan banyak lainnya

In [ ]:
# max dan min temp selama 140 hari terakhir berapa?
print("max", arr_temp.max())
print("min", arr_temp.min())

# max dan min per minggu?
print("max mingguan", arr_temp.max(axis=1))
print("min mingguan", arr_temp.min(axis=1))

max 44.7
min 15.1
max mingguan [37.6 43.2 42.9 38.6 38.7 44.7 41.3 42.3 44.  37.5 44.  39.6 42.6 41.1
 40.2 44.3 42.3 41.7 43.1 40.3]
min mingguan [19.4 16.8 17.7 20.  18.1 28.5 16.3 16.6 15.9 17.1 15.1 19.7 16.  15.9
 16.8 26.7 20.5 18.6 18.2 22. ]


kita bisa cobain juga aritmatiknya

In [ ]:
arr = np.arange(12)
arr + 3 - 5 * 4 / 2

array([-7., -6., -5., -4., -3., -2., -1.,  0.,  1.,  2.,  3.,  4.])

In [ ]:
arr2 = np.arange(12,24)
arr2.shape

(12,)

In [ ]:
# bukan concatenate nempel, tapi angkanya yang ditambah
arr + arr2

array([12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34])

# Matrix
numpy array 2 dimensi


## dot product
cara spesial mengkalikan 2 matrix

In [ ]:

arr_penjualan = np.array([[5,10,3],
                          [6,12,7],
                          [5,13,1],
                          [10,11,20]])

arr_profit = np.array([ [2000],
                        [1000],
                        [1500]])

print(arr_penjualan)
print(arr_profit)

[[ 5 10  3]
 [ 6 12  7]
 [ 5 13  1]
 [10 11 20]]
[[2000]
 [1000]
 [1500]]


In [ ]:
profit_harian = arr_penjualan.dot(arr_profit)
print(profit_harian)

[[24500]
 [34500]
 [24500]
 [61000]]


In [ ]:
profit_total = profit_harian.sum()
print(profit_total)

144500


## Tranpose dan Reshape
transpose berbeda dengan reshape

transpose menyimpan teman sebaris menjadi teman sekolom

reshape tidak peduli dengan teman sebaris/sekolom

In [ ]:
# misal kita punya data penjualan harian
print(arr_penjualan)

[[ 5 10  3]
 [ 6 12  7]
 [ 5 13  1]
 [10 11 20]]


setiap baris adalah hari berbeda

setiap kolom adalah produk berbeda

kalau kita pakai reshape

In [ ]:
arr_penjualan.reshape((3,4))

array([[ 5, 10,  3,  6],
       [12,  7,  5, 13],
       [ 1, 10, 11, 20]])

datanya tercampur. perhatikan penjualan senin 5 10 3 menjadi 5 12 1 di reshape sini. ini karena reshape tidak peduli dengan posisi per baris atau per kolom.

jadi biasanya kita akan menggunakan transpose kalau datanya data riil.

In [ ]:
arr_penjualan.T

array([[ 5,  6,  5, 10],
       [10, 12, 13, 11],
       [ 3,  7,  1, 20]])

transpose memastikan yang tadinya baris jadi kolom, dan yang tadinya kolom jadi baris

### latihan pendek

- gunakan numpy array untuk membuat dummy data hasil lempar dadu, dari antara angka 1-6
- anggaplah ada 3 orang yang sedang bermain dadu, masing masing melempar 10 kali
- hitung berapa rata rata nilai dadu yang dimiliki per orang
- hitung total nilai yang dimiliki per orang
- siapa yang memiliki nilai tertinggi (orang ke berapa/ baris ke berapa)

In [ ]:
dummy_data = np.round(np.random.random((3,10))*5+1, 0)
print("dummy data")
print(dummy_data)

dummy data
[[1. 3. 3. 4. 5. 2. 3. 3. 5. 5.]
 [3. 4. 2. 6. 3. 4. 2. 6. 3. 2.]
 [2. 5. 6. 3. 4. 1. 3. 1. 3. 3.]]


In [ ]:
print("rata2 per orang")
mean_dd = dummy_data.mean(axis=1)
print(mean_dd)

rata2 per orang
[3.4 3.5 3.1]


In [ ]:
print("total per orang")
sum_dd = dummy_data.sum(axis=1)
print(sum_dd)

total per orang
[34. 35. 31.]


di kasus ini, orang ke2 memiliki nilai tertinggi di 35

In [ ]:
highest_per_person = sum_dd.max()
print(highest_per_person)

35.0


In [ ]:
# kita bisa gunakan cara lebih otomatis juga untuk mencari index
# baris ke berapa yang punya total nilai tertinggi
np.where(sum_dd == highest_per_person)[0][0]

1

# Preview: Pandas
package pandas memiliki dua data type
- Series - 1 dimensi
- DataFrame - 2 dimensi

berbeda sama numpy, yang sama sama numpy array mau 1D, 2D, 3D, ...

di pandas cuma ada 2 pilihan: 1D Series, atau 2D DataFrame

untuk contoh kali ini, kita akan gunakan DataFrame

In [ ]:
import pandas as pd

In [ ]:
# bikin dataframe
pd.DataFrame(dummy_data)

0    1    2    3    4    5    6    7    8    9
0  1.0  3.0  3.0  4.0  5.0  2.0  3.0  3.0  5.0  5.0
1  3.0  4.0  2.0  6.0  3.0  4.0  2.0  6.0  3.0  2.0
2  2.0  5.0  6.0  3.0  4.0  1.0  3.0  1.0  3.0  3.0

In [ ]:
col_names = []
for n in range(1, 11):
  col_names.append("Trial " + str(n))
col_names

['Trial 1',
 'Trial 2',
 'Trial 3',
 'Trial 4',
 'Trial 5',
 'Trial 6',
 'Trial 7',
 'Trial 8',
 'Trial 9',
 'Trial 10']

In [ ]:
# di dataframe, kita bisa kasih judul kolom dan judul baris

pd.DataFrame(dummy_data, columns=col_names)

Trial 1  Trial 2  Trial 3  Trial 4  Trial 5  Trial 6  Trial 7  Trial 8  \
0      1.0      3.0      3.0      4.0      5.0      2.0      3.0      3.0   
1      3.0      4.0      2.0      6.0      3.0      4.0      2.0      6.0   
2      2.0      5.0      6.0      3.0      4.0      1.0      3.0      1.0   

   Trial 9  Trial 10  
0      5.0       5.0  
1      3.0       2.0  
2      3.0       3.0

In [ ]:
# judul baris namanya index

pd.DataFrame(dummy_data, columns=col_names, index=["Andi", "Budi", "Calvin"])

Trial 1  Trial 2  Trial 3  Trial 4  Trial 5  Trial 6  Trial 7  \
Andi        1.0      3.0      3.0      4.0      5.0      2.0      3.0   
Budi        3.0      4.0      2.0      6.0      3.0      4.0      2.0   
Calvin      2.0      5.0      6.0      3.0      4.0      1.0      3.0   

        Trial 8  Trial 9  Trial 10  
Andi        3.0      5.0       5.0  
Budi        6.0      3.0       2.0  
Calvin      1.0      3.0       3.0

## latihan pendek
- buatlah data tabular 2 dimensi menggunakan numpy random random
- masukkan data tersebut ke sebuah pandas DataFrame
- berikan judul kolom dan judul baris

In [ ]:
pd.DataFrame([[1,2,3], [4,5,6]], columns=["A", "B", "C"], index=["I", "II"])

A  B  C
I   1  2  3
II  4  5  6

In [ ]:
# misalkan kita buat data penjualan alat tulis
# ada pensil, penghapus, buku tulis
# kita jual, ada kolom jumlah terjual, dan kolom harga per item

judul_kolom = ["Jumlah Terjual", "Harga per Item"]
judul_baris = ["Pensil", "Penghapus", "Buku Tulis"]

# kita perlu data 3 baris 2 kolom
# kita gunakan numpy random random
# kita random jumlah terjual dulu, baru harga per item

jumlah_terjual = np.round(np.random.random((3, 1)) * 50 + 50, 0)
harga_per_item = np.array([5000, 3500, 15000]).reshape((3,1))

data = np.concatenate((jumlah_terjual,harga_per_item), axis=1)
data

array([[   66.,  5000.],
       [   71.,  3500.],
       [   90., 15000.]])

In [ ]:
df = pd.DataFrame(data, columns=judul_kolom, index=judul_baris)
df

Jumlah Terjual  Harga per Item
Pensil                78.0          5000.0
Penghapus             92.0          3500.0
Buku Tulis            72.0         15000.0